File: src/main/com/ypwang/medium/leetcode406.java
package com.ypwang.medium;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

public class leetcode406 {
    public int[][] reconstructQueue(int[][] people) {
        Arrays.sort(people, (a, b) -> a[0] == b[0] ? a[1] - b[1] : b[0] - a[0]);
        List<int[]> list = new LinkedList<>();
        for (int i = 0; i < people.length; i++) {
            list.add(people[i][1], people[i]);
        }
        int[][] res = new int[people.length][2];
        for (int i = 0; i < res.length; i++) {
            res[i] = list.get(i);
        }
        return res;
    }
}

File: src/main/com/ypwang/medium/leetcode406.kt
package com.ypwang.medium

class Solution406 {
    fun reconstructQueue(people: Array<IntArray>): Array<IntArray> {
        val p = people.sortedWith(
            Comparator { a1, a2 ->
                if (a1[0] == a2[0])
                    a1[1] - a2[1]
                else
                    a2[0] - a1[0]
            }
        )
        val rst = mutableListOf<IntArray>()
        for (q in p) {
            rst.add(q[1], q)
        }
        return rst.toTypedArray()
    }
}

fun main() {
    println(Solution406().reconstructQueue(
            arrayOf(
                    intArrayOf(7, 0), intArrayOf(4, 4), intArrayOf(7, 1), intArrayOf(5, 0), intArrayOf(6, 1),
                    intArrayOf(5, 2)
            )
    ).map { it.toList() })
}